README
INSTALLATION:

Thank you for trying duplicity.  To install, run:

	python setup.py install

The build process can be also be run separately:

	python setup.py build

The default prefix is /usr, so files are put in /usr/bin,
/usr/share/man/, etc.  An alternate prefix can be specified using the
--prefix=<prefix> option.  For example:

	python setup.py install --prefix=/usr/local


REQUIREMENTS:

Remember that you must have Python 2.2 or later and librsync 0.9.5.1
or later installed.  For Python, see http://www.python.org.  The
rdiff-backup homepage at http://rdiff-backup.stanford.edu/ should have
a recent version of librsync; otherwise see the librsync homepage at
http://sourceforge.net/projects/librsync/.



CHANGELOG
New in v0.6.19 (2012/05/22)
---------------------------
Enhancements:
- lots of work on the man page to clean up requirements, etc.
- use empty listbody for enhanced webdav compatibility
- initial folder creation on backend does not result in a ResponseNotReady anymore
- add ssh_config support (/etc/ssh/ssh_config + ~/.ssh/config) to paramiko sshbackend
- add missing_host_key prompt to new sshbackend similar to ssh procedure
- added --ssh-backend parameter to switch between paramiko,pexpect
- allow answering gio mount questions (albeit naively)
- if the gio backend wants to ask a question during its mount phase, it previously just aborted.
- a couple more warning error codes that Deja Dup is interested in noticing.
- ssh paramiko backend respects --num-retries now
- set retry delay for ssh backends to 10s
- ssh pexpect backend
 + sftp part does not claim 'Invalid SSH password' although it's only 'Permission denied' now
 + sftp errors are now more talkative
- gpg.py
 + commented assert which broke otherwise working verify run

Bugs closed in this release:
588541     Connection failed, please check your password: Login dialog cancelled

Merges:
lp:~ed.so/duplicity/0.6-manpage
lp:~ed.so/duplicity/0.6-webdav_fixes
lp:~carlos-abalde/duplicity/gdocs-backend-gdata-2.0.16.-upgrade.
lp:~ed.so/duplicity/0.6-ssh_add_missinghostkey
lp:~ed.so/duplicity/0.6-readd_sshpexpect


New in v0.6.18 (2012/02/29)
---------------------------
Enhancements:
- fix extraneous '.py' in botobackend.py include
- tests: add delay between backups to avoid assertion error
- tests: use backup source that is more likely to be larger than 1M compressed
- tests: make other-filesystem check more robust against certain directories being mounts or not
- resuming an incremental results in a 'Restarting backup, but current encryption settings do not match original settings' error because curtime is incorrectly set away from previous incremental value
- added option to not compress the backup, when no encryption is selected
- always delay a little bit when a backend gives us errors
- Don't cache TarInfo files. Tests still pass, so I don't believe we need the members cache (and in the old tarfile.py, we didn't cache either).
- Adding --file-prefix option so different sets of backups can be stored in the same bucket.  See blueprint at https://blueprints.launchpad.net/duplicity/+spec/file-prefix-option
- two changes that help the test suite pass
- raise log level on backend import failure so it will be visible under default conditions
- file /etc/motd may not exist in test environment.  Use __file__ instead to point to a known plaintext source file.
- some code/import changes to make the ssh and boto backends compatible with Python 2.4.
- some changes to make roottest.py compatible with the new dir structure.

Bugs closed in this release:
884638     Python 2.5 / boto error
908228     possible memory leak
909031     SSH-Backend: Creating dirs separately causes a permissons-problems
916689     multipart upload fails on python 2.7.2
929465     UnsupportedBackendScheme: scheme not supported in url: scp://u123@u123.example.com/foo/
930727     ftpsbackend should respect num_retries for ftp commands
931175     duplicity crashes when PYTHONOPTIMIZE is set

Merges:
lp:~mterry/duplicity/always-delay
lp:~mterry/duplicity/memleak
lp:~mterry/duplicity/nopexpect
lp:~mterry/duplicity/resume-inc
lp:~mterry/duplicity/testfixes
lp:~nguyenqmai/duplicity/file-prefix-option
lp:~tobias-genannt/duplicity/nocompress


New in v0.6.17 (2011/11/25)
---------------------------
Enhancements:
- Added --rsync-options flag to allow user to pass options to rsync at will
- Added --s3-use-multiprocessing to select the new s3 multiprocessing backend.
  Default is to use the single processing backend.  A helper, filechunkio.py,
  requires Python 2.6+, so this option is not usable in earlier versions.

Bugs closed in this release:
411145     Misleading error message: "Invalid SSH password"
871875     File ... was corrupted during upload.
878220	   UnsupportedBackendScheme: scheme not supported in url: s3+http
878964     Resuming a backup with a different password should throw an error
881070	   Bugfix for S3 multipart upload on 0.6.16
881727	   duplicity 0.6.16 lists all backup sets as incomplete
885670	   Duplicity fails to add incremental backups to chain...

Merges:
lp:~mterry/duplicity/check-passphrase-on-restart
lp:~moss-d/duplicity/rsync-options


New in v0.6.16 (2011/10/16)
----------------------------
Enhancements:
- Usability enhancement: sign passphrase prompt has no second
  verification prompt anymore, symmetric passphrases are
  still verified
- Fixed Unicode errors when translations are used.
- Replaced old tarfile.py with Python 2.7 version, modded
  to support Python 2.4 syntax.

Bugs closed in this release:
485219     Fixed fields are not fixed, leading to buffer overflows...
676109     Amazon S3 backend multipart upload support
690549     uid and guid setting adversely affects integrity
739438     Local backend should always try renaming instead of copying
832149     Uploads to Rackspace fail silently
835892     duplicity crash: "AssertionError: rb None None"
838162     Duplicity URL Parser is not parsing IPv6 properly
838264     Duplicity thinks partial encrypted backups are not encrypted
870116     Duplicity does not handle UIDs higher than 60001

Merges:
lp:~ed.so/duplicity/reuse-passphrase-for-signing-fix
lp:~mterry/duplicity/cloudfiles-10k
lp:~duplicity-team/duplicity/check-volumes
lp:~mterry/duplicity/tarfile
lp:~mterry/duplicity/partial-encryption
lp:~mterry/duplicity/fix-local-backend-validation
lp:~ross-ross-williams/duplicity/gpg-agent-fix
lp:~mterry/duplicity/rbNoneNone
lp:~ed.so/duplicity/UnicodeDecodeError


New in v0.6.15 (2011/08/19)
---------------------------
Enhancements:
- Ignore 404 errors when deleting a file on Ubuntu One.
- Ignore ENOENT (file missing) errors where it is safe.
- Set minimum Python version to 2.4 in README.

src/log.py
# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
#
# Copyright 2002 Ben Escoto <ben@emerose.org>
# Copyright 2007 Kenneth Loafman <kenneth@loafman.com>
# Copyright 2008 Michael Terry <mike@mterry.name>
#
# This file is part of duplicity.
#
# Duplicity is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# Duplicity is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with duplicity; if not, write to the Free Software Foundation,
# Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

"""Log various messages depending on verbosity level"""

import os
import sys
import logging

MIN = 0
ERROR = 0
WARNING = 2
NOTICE = 3
INFO = 5
DEBUG = 9
MAX = 9

_logger = None

def DupToLoggerLevel(verb):
    """Convert duplicity level to the logging module's system, where higher is
       more severe"""
    return MAX - verb + 1

def LoggerToDupLevel(verb):
    """Convert logging module level to duplicity's system, where lowere is
       more severe"""
    return DupToLoggerLevel(verb)

def Log(s, verb_level, code=1, extra=None, force_print=False):
    """Write s to stderr if verbosity level low enough"""
    global _logger
    # controlLine is a terrible hack until duplicity depends on Python 2.5
    # and its logging 'extra' keyword that allows a custom record dictionary.
    if extra:
        _logger.controlLine = '%d %s' % (code, extra)
    else:
        _logger.controlLine = '%d' % (code)
    if not s:
        s = '' # If None is passed, standard logging would render it as 'None'

    if force_print:
        initial_level = _logger.getEffectiveLevel()
        _logger.setLevel(DupToLoggerLevel(MAX))

    _logger.log(DupToLoggerLevel(verb_level), s.decode("utf8", "ignore"))
    _logger.controlLine = None

    if force_print:
        _logger.setLevel(initial_level)

def Debug(s):
    """Shortcut used for debug message (verbosity 9)."""
    Log(s, DEBUG)

class InfoCode:
    """Enumeration class to hold info code values.
       These values should never change, as frontends rely upon them.
       Don't use 0 or negative numbers."""
    generic = 1
    progress = 2
    collection_status = 3
    diff_file_new = 4
    diff_file_changed = 5
    diff_file_deleted = 6
    patch_file_writing = 7
    patch_file_patching = 8
    #file_list = 9 # 9 isn't used anymore.  It corresponds to an older syntax for listing files
    file_list = 10

def Info(s, code=InfoCode.generic, extra=None):
    """Shortcut used for info messages (verbosity 5)."""
    Log(s, INFO, code, extra)

def Progress(s, current, total=None):
    """Shortcut used for progress messages (verbosity 5)."""
    if total:
        controlLine = '%d %d' % (current, total)
    else:
        controlLine = '%d' % current
    Log(s, INFO, InfoCode.progress, controlLine)

def PrintCollectionStatus(col_stats, force_print=False):
    """Prints a collection status to the log"""
    Log(str(col_stats), 8, InfoCode.collection_status,
        '\n' + '\n'.join(col_stats.to_log_info()), force_print)

def Notice(s):
    """Shortcut used for notice messages (verbosity 3, the default)."""
    Log(s, NOTICE)

class WarningCode:
    """Enumeration class to hold warning code values.
       These values should never change, as frontends rely upon them.
       Don't use 0 or negative numbers."""
    generic = 1
    orphaned_sig = 2
    unnecessary_sig = 3
    unmatched_sig = 4
    incomplete_backup = 5
    orphaned_backup = 6
    ftp_ncftp_v320 = 7 # moved from error

def Warn(s, code=WarningCode.generic):
    """Shortcut used for warning messages (verbosity 2)"""
    Log(s, WARNING, code)

class ErrorCode:
    """Enumeration class to hold error code values.
       These values should never change, as frontends rely upon them.
       Don't use 0 or negative numbers.  This code is returned by duplicity
       to indicate which error occurred via both exit code and log."""
    generic = 1 # Don't use if possible, please create a new code and use it
    command_line = 2
    source_mismatch = 3
    no_manifests = 4
    mismatched_manifests = 5
    unreadable_manifests = 6
    cant_open_filelist = 7
    bad_url = 8
    bad_archive_dir = 9
    bad_sign_key = 10
    restore_dir_exists = 11
    verify_dir_doesnt_exist = 12
    backup_dir_doesnt_exist = 13
    file_prefix_error = 14
    globbing_error = 15
    redundant_inclusion = 16
    inc_without_sigs = 17
    no_sigs = 18
    restore_dir_not_found = 19
    no_restore_files = 20
    mismatched_hash = 21
    unsigned_volume = 22
    user_error = 23
    boto_old_style = 24
    boto_lib_too_old = 25
    boto_calling_format = 26
    ftp_ncftp_missing = 27
    ftp_ncftp_too_old = 28
    #ftp_ncftp_v320 = 29 # moved to warning
    exception = 30
    gpg_failed = 31
    s3_bucket_not_style = 32
    not_implemented = 33
    get_freespace_failed = 34
    not_enough_freespace = 35
    get_ulimit_failed = 36
    maxopen_too_low = 37
    connection_failed = 38

def FatalError(s, code, extra=None):
    """Write fatal error message and exit"""
    Log(s, ERROR, code, extra)
    shutdown()
    sys.exit(code)

class DupLogRecord(logging.LogRecord):
    """Custom log record that holds a message code"""
    def __init__(self, controlLine, *args, **kwargs):
        global _logger
        logging.LogRecord.__init__(self, *args, **kwargs)
        self.controlLine = controlLine

class DupLogger(logging.Logger):
    """Custom logger that creates special code-bearing records"""
    # controlLine is a terrible hack until duplicity depends on Python 2.5
    # and its logging 'extra' keyword that allows a custom record dictionary.
    controlLine = None
    def makeRecord(self, name, lvl, fn, lno, msg, args, exc_info, *argv, **kwargs):
        return DupLogRecord(self.controlLine, name, lvl, fn, lno, msg, args, exc_info)

class OutFilter(logging.Filter):
    """Filter that only allows warning or less important messages"""
    def filter(self, record):
        return record.msg and record.levelno <= DupToLoggerLevel(WARNING)

class ErrFilter(logging.Filter):
    """Filter that only allows messages more important than warnings"""
    def filter(self, record):
        return record.msg and record.levelno > DupToLoggerLevel(WARNING)

def setup():
    """Initialize logging"""
    global _logger
    if _logger:
        return

    logging.setLoggerClass(DupLogger)
    _logger = logging.getLogger("duplicity")

    # Set up our special level names
    logging.addLevelName(DupToLoggerLevel(0), "ERROR")
    logging.addLevelName(DupToLoggerLevel(1), "WARNING")
    logging.addLevelName(DupToLoggerLevel(2), "WARNING")
    logging.addLevelName(DupToLoggerLevel(3), "NOTICE")
    logging.addLevelName(DupToLoggerLevel(4), "NOTICE")
    logging.addLevelName(DupToLoggerLevel(5), "INFO")
    logging.addLevelName(DupToLoggerLevel(6), "INFO")
    logging.addLevelName(DupToLoggerLevel(7), "INFO")
    logging.addLevelName(DupToLoggerLevel(8), "INFO")
    logging.addLevelName(DupToLoggerLevel(9), "DEBUG")

    # Default verbosity allows notices and above
    setverbosity(NOTICE)

    # stdout and stderr are for different logging levels
    outHandler = logging.StreamHandler(sys.stdout)
    outHandler.addFilter(OutFilter())
    _logger.addHandler(outHandler)

    errHandler = logging.StreamHandler(sys.stderr)
    errHandler.addFilter(ErrFilter())
    _logger.addHandler(errHandler)

class MachineFormatter(logging.Formatter):
    """Formatter that creates messages in a syntax easily consumable by other
       processes."""
    def __init__(self):
        # 'message' will be appended by format()
        logging.Formatter.__init__(self, "%(levelname)s %(controlLine)s")

    def format(self, record):
        s = logging.Formatter.format(self, record)

        # Add user-text hint of 'message' back in, with each line prefixed by a
        # dot, so consumers know it's not part of 'controlLine'
        if record.message:
            s += ('\n' + record.message).replace('\n', '\n. ')

        # Add a newline so consumers know the message is over.
        return s + '\n'

class MachineFilter(logging.Filter):
    """Filter that only allows levels that are consumable by other processes."""
    def filter(self, record):
        # We only want to allow records that have level names.  If the level
        # does not have an associated name, there will be a space as the
        # logging module expands levelname to "Level %d".  This will confuse
        # consumers.  Even if we dropped the space, random levels may not
        # mean anything to consumers.
        s = logging.getLevelName(record.levelno)
        return s.find(' ') == -1

def add_fd(fd):
    """Add stream to which to write machine-readable logging"""
    global _logger
    handler = logging.StreamHandler(os.fdopen(fd, 'w'))
    handler.setFormatter(MachineFormatter())
    handler.addFilter(MachineFilter())
    _logger.addHandler(handler)

def add_file(filename):
    """Add file to which to write machine-readable logging"""
    global _logger
    handler = logging.FileHandler(filename, 'w')
    handler.setFormatter(MachineFormatter())
    handler.addFilter(MachineFilter())
    _logger.addHandler(handler)

def setverbosity(verb):
    """Set the verbosity level"""
    global _logger
    _logger.setLevel(DupToLoggerLevel(verb))

def getverbosity():
    """Get the verbosity level"""
    global _logger
    return LoggerToDupLevel(_logger.getEffectiveLevel())

def shutdown():
    """Cleanup and flush loggers"""
    logging.shutdown()

